plugins/aggregators/derivative/derivative.go

```diff
@@ -0,0 +1,225 @@
+	# variable = ""
+	##
+	## When using a field as the derivation parameter the name of that field will
+	## be used for the resulting derivative, e.g. *fieldname_by_parameter*.
+	##
+	## Note, that the calculation is based on the actual timestamp of the
+	## measurements. When there is only one measurement during that period, the
+	## measurement will be rolled over to the next period. The maximum number of
+	## such roll-overs can be configured with a default of 10.
+	# max_roll_over = 10
+	##
+`
+
+func (d *Derivative) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *Derivative) Description() string {
+	return "Calculates a derivative for every field."
+}
+
+func (d *Derivative) Add(in telegraf.Metric) {
+	id := in.HashID()
+	if current, ok := d.cache[id]; !ok {
+		// hit an uncached metric, create caches for first time:
+		d.cache[id] = newAggregate(in)
+	} else {
+		if current.first.time.After(in.Time()) {
+			current.first = newEvent(in)
+			current.rollOver = 0
+		} else if current.first.time.Equal(in.Time()) {
+			upsertConvertedFields(in.Fields(), current.first.fields)
+			current.rollOver = 0
+		}
+		if current.last.time.Before(in.Time()) {
+			current.last = newEvent(in)
+			current.rollOver = 0
+		} else if current.last.time.Equal(in.Time()) {
+			upsertConvertedFields(in.Fields(), current.last.fields)
+			current.rollOver = 0
+		}
+	}
+}
+
+func newAggregate(in telegraf.Metric) *aggregate {
+	event := newEvent(in)
+	return &aggregate{
+		name:     in.Name(),
+		tags:     in.Tags(),
+		first:    event,
+		last:     event,
+		rollOver: 0,
+	}
+}
+
+func newEvent(in telegraf.Metric) *event {
+	return &event{
+		fields: extractConvertedFields(in),
+		time:   in.Time(),
+	}
+}
+
+func extractConvertedFields(in telegraf.Metric) map[string]float64 {
+	fields := make(map[string]float64, len(in.Fields()))
+	upsertConvertedFields(in.Fields(), fields)
+	return fields
+}
+
+func upsertConvertedFields(source map[string]interface{}, target map[string]float64) {
+	for k, v := range source {
+		if value, ok := convert(v); ok {
+			target[k] = value
+		}
+	}
+}
+
+func convert(in interface{}) (float64, bool) {
+	switch v := in.(type) {
+	case float64:
+		return v, true
+	case int64:
+		return float64(v), true
+	case uint64:
+		return float64(v), true
+	}
+	return 0, false
+}
+
+func (d *Derivative) Push(acc telegraf.Accumulator) {
+	for _, aggregate := range d.cache {
+		if aggregate.first == aggregate.last {
+			d.Log.Warnf("Same first and last event for %q, skipping.", aggregate.name)
+			continue
+		}
+		var denominator float64
+		if len(d.Variable) == 0 {
+			// If no derivation variable is known default to timestamp
+			denominator = aggregate.last.time.Sub(aggregate.first.time).Seconds()
+		} else {
+			var first float64
+			var last float64
+			var found bool
+			if first, found = aggregate.first.fields[d.Variable]; !found {
+				d.Log.Warnf("Did not find %q in first event for %q.", d.Variable, aggregate.name)
```

Comment:
Might want to move these to debug log statements. I don't think they're that useful for normal operation.

Reply:
It was a suggestion by @srebhan to log at warning level, whenever no value would be emitted. Logging the different reasons already proved valuable to me during debugging the plugin.

Reply:
@ssoroka my reasoning was to allow the user to find out why there was no metric emitted. We could change the log-level to debug if you fear too many log-messages... In general this should only trigger if you get a new _field_ key in an interval that was not seen before (something that should not happen IIRC).

Reply:
Do I need to change something here?

Reply:
@ssoroka?

Reply:
yes, use debug please.

Reply:
```suggestion
				d.Log.Debugf("Did not find %q in first event for %q.", d.Variable, aggregate.name)
```